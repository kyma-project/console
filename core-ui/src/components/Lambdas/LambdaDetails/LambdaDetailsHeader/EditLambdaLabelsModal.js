import React, { useState, useEffect } from 'react';

import { Icon } from 'fundamental-react';
import { Tooltip } from 'react-shared';

import ModalWithForm from 'components/ModalWithForm/ModalWithForm';
import { LabelsInput } from 'components/Lambdas/components';

import { useUpdateLambda, UPDATE_TYPE } from 'components/Lambdas/gql/hooks';

import { LAMBDA_DETAILS } from 'components/Lambdas/constants';

function EditLambdaLabelsForm({
  lambda,
  formElementRef,
  setValidity = () => void 0,
}) {
  const updateLambda = useUpdateLambda({
    lambda,
    type: UPDATE_TYPE.GENERAL_CONFIGURATION,
  });
  const [labels, setLabels] = useState(lambda.labels);

  useEffect(() => {
    setValidity(true);
  }, [setValidity]);

  function updateLabels(newLabels) {
    setLabels(newLabels);
  }

  async function handleSubmit() {
    updateLambda({
      labels,
    });
  }

  return (
    <form
      ref={formElementRef}
      onSubmit={handleSubmit}
      className="lambda-details-header__edit-labels-form"
    >
      <LabelsInput labels={labels} onChange={updateLabels} />
    </form>
  );
}

export default function EditLambdaLabelsModal({ lambda }) {
  const modalOpeningComponent = (
    <Tooltip
      title={LAMBDA_DETAILS.LABELS.POPUP_MESSAGE}
      position="top"
      trigger="mouseenter"
      tippyProps={{
        distance: 8,
      }}
    >
      <Icon glyph="edit" />
    </Tooltip>
  );

  return (
    <ModalWithForm
      title={LAMBDA_DETAILS.LABELS.EDIT_MODAL.TITLE}
      modalOpeningComponent={modalOpeningComponent}
      confirmText={LAMBDA_DETAILS.LABELS.EDIT_MODAL.CONFIRM_BUTTON.TEXT}
      id="edit-labels-modal"
      className="lambda-details-header__edit-labels-modal"
      renderForm={props => <EditLambdaLabelsForm lambda={lambda} {...props} />}
    />
  );
}
