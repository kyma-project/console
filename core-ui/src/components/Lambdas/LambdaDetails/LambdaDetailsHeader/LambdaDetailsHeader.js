import React from 'react';
import PropTypes from 'prop-types';
import { ActionBar, Button, Breadcrumb } from 'fundamental-react';
import LuigiClient from '@kyma-project/luigi-client';
import { useMutation } from '@apollo/react-hooks';

import { handleDelete } from 'react-shared';
import { DELETE_LAMBDA } from '../../../../gql/mutations';
import { useNotification } from '../../../../contexts/notifications';
import { REFETCH_TIMEOUT } from '../../../../shared/constants';

LambdaDetailsHeader.propTypes = {
  lambda: PropTypes.object.isRequired,
  handleUpdate: PropTypes.func.isRequired,
};

export default function LambdaDetailsHeader({ lambda, handleUpdate }) {
  const { name, namespace } = lambda;
  const [deleteLambda] = useMutation(DELETE_LAMBDA);
  const notificationManager = useNotification();

  const handleLambdaDelete = async () => {
    try {
      const deletedLambda = await deleteLambda({
        variables: { name, namespace },
      });
      const isSuccess =
        deletedLambda.data &&
        deletedLambda.data.deleteFunction &&
        deletedLambda.data.deleteFunction.name === name;
      if (isSuccess) {
        notificationManager.notify({
          content: `Lambda ${name} deleted`,
          title: 'Success',
          color: '#107E3E',
          icon: 'accept',
          autoClose: true,
        });
      }
    } catch (e) {
      notificationManager.notify({
        content: `Error while removing lambda ${name}: ${e.message}`,
        title: 'Error',
        color: '#BB0000',
        icon: 'decline',
        autoClose: false,
      });
    }
  };

  const navigateToList = () => {
    LuigiClient.linkManager()
      .fromClosestContext()
      .navigate('');
  };

  return (
    <header className="fd-has-background-color-background-2">
      <section className="fd-has-padding-regular fd-has-padding-bottom-none action-bar-wrapper">
        <section>
          <Breadcrumb>
            <Breadcrumb.Item
              name="Lambdas"
              url="#"
              onClick={() => navigateToList()}
            />
            <Breadcrumb.Item />
          </Breadcrumb>
          <ActionBar.Header title={name || 'Loading name...'} />
        </section>
        <ActionBar.Actions>
          <Button onClick={handleUpdate} data-test-id="lambda-save-button">
            Save
          </Button>
          <Button
            onClick={() => {
              handleDelete('Lambda', name, name, handleLambdaDelete, () => {
                setTimeout(() => {
                  navigateToList();
                }, REFETCH_TIMEOUT);
              });
            }}
            option="light"
            type="negative"
          >
            Delete
          </Button>
        </ActionBar.Actions>
      </section>
    </header>
  );
}
