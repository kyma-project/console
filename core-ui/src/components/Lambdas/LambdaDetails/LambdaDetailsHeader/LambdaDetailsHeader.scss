.lambda-details-header {
  .fd-panel__header {
    padding-right: 32px;
    padding-left: 32px;
    padding-bottom: 8px;
    border-bottom-width: 0;

    .fd-has-margin-bottom-s {
      margin-bottom: 0px !important;
    }
  }

  &__content {
    background: #fff;
    padding: 0 32px;
    margin-bottom: -32px;
  }

  &__status {
    .fd-panel {
      border-width: 1px 1px 1px 6px;
      border-style: solid;
      border-color: rgb(217, 217, 217);
      border-image: initial;
      box-shadow: none;

      .fd-panel__header {
        border-bottom-width: 1px;
      }
    }

    &--success {
      .fd-panel {
        border-left-color: #3db350;
      }
      .lambda-details-header__status-phase {
        color: #3db350;
      }
    }

    &--warning {
      .fd-panel {
        border-left-color: rgb(233, 115, 12);
      }
      .lambda-details-header__status-phase {
        color: rgb(233, 115, 12);
      }
      .sap-icon--sys-help {
        color: rgb(233, 115, 12);
      }
    }

    &--error {
      .fd-panel {
        border-left-color: #ee0000;
      }
      .lambda-details-header__status-phase {
        color: #ee0000;
      }
    }

    &--info {
      .fd-panel {
        border-left-color: rgb(10, 110, 209);
      }
      .lambda-details-header__status-phase {
        color: rgb(10, 110, 209);
      }
      .sap-icon--sys-help {
        color: rgb(10, 110, 209);
      }
    }

    .sap-icon--sys-enter {
      color: #3db350;
    }

    .sap-icon--sys-cancel {
      color: #ee0000;
    }

    .fd-panel__header {
      color: rgb(50, 54, 58);
      font-size: 16px;
      font-weight: normal;
      padding: 16px 24px;
    }

    &-phase {
      text-transform: uppercase;
      font-weight: 700;
      margin-bottom: 0;
    }

    &-message {
      margin-top: 6px;
    }

    .link {
      margin-left: 6px;
    }
  }

  &__labels {
    display: inline-block;

    &-title,
    &-content {
      display: block;
    }
  
    &-title {
      color: var(--fd-color-text-4);
      margin-bottom: 6px;
      font-size: 14px;

      > div {
        display: inline-block;
      }

      .sap-icon--edit {
        padding-left: 8px;
        cursor: pointer;

        &:hover {
          color: rgb(10, 110, 209);
        } 
      }
    }

    .fd-token {
      margin: 0 8px 8px 0;
    }
  }

  &__edit-labels-modal {
    .fd-modal__body {
      min-height: 140px;
    }
  }
}