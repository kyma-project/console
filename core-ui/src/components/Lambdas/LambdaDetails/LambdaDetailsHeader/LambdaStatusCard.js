import React, { useState } from 'react';
import LuigiClient from '@kyma-project/luigi-client';

import { Panel, Icon, Modal, Button } from 'fundamental-react';

import { statusType } from 'components/Lambdas/helpers/lambdas';
import {
  LAMBDA_PHASES,
  LAMBDA_ERROR_PHASES,
  LAMBDA_DETAILS,
  BUTTONS,
} from 'components/Lambdas/constants';
import { formatMessage } from 'components/Lambdas/helpers/misc';

function statusIcon(phase = LAMBDA_PHASES.INITIALIZING.TYPE) {
  switch (phase) {
    case LAMBDA_PHASES.FAILED.TYPE:
      return 'sys-cancel';
    case LAMBDA_PHASES.RUNNING.TYPE:
      return 'sys-enter';
    default:
      return 'sys-help';
  }
}

function ModalWithErrorLogs({ lambdaName, error }) {
  const [show, setShow] = useState(false);

  function onOpen() {
    setShow(true);
    LuigiClient.uxManager().addBackdrop();
  }

  function onClose() {
    setShow(false);
    LuigiClient.uxManager().removeBackdrop();
  }

  const title = formatMessage(LAMBDA_DETAILS.STATUS.ERROR.MODAL.TITLE, {
    lambdaName,
  });
  const actions = (
    <Button onClick={onClose} option="light">
      {BUTTONS.CANCEL}
    </Button>
  );

  return (
    <>
      <span className="link" onClick={onOpen}>
        {LAMBDA_DETAILS.STATUS.ERROR.LINK}
      </span>
      <Modal title={title} show={show} actions={actions} onClose={onClose}>
        {error}
      </Modal>
    </>
  );
}

export default function LambdaStatusCard({ lambdaName, status }) {
  const statusPhase = status.phase;

  const texts = LAMBDA_PHASES[statusPhase];
  const className = statusType(statusPhase);
  const glyph = statusIcon(statusPhase);

  // for Failed and NewRevisionError phases error message is show in modal
  const hasErrorPhase = LAMBDA_ERROR_PHASES.includes(statusPhase);
  let message = null;
  if (hasErrorPhase) {
    const errorLogs = (
      <ModalWithErrorLogs lambdaName={lambdaName} error={status.message} />
    );

    message = (
      <p className="lambda-details-header__status-message">
        <span>{texts.MESSAGE}</span>
        {errorLogs}
      </p>
    );
  }

  return (
    <div
      className={`lambda-details-header__status lambda-details-header__status--${className}`}
    >
      <Panel>
        <Panel.Header>
          <Panel.Head title={LAMBDA_DETAILS.STATUS.TITLE} />
          <Panel.Actions>
            <Icon glyph={glyph} />
          </Panel.Actions>
        </Panel.Header>
        <Panel.Body>
          <h3 className="lambda-details-header__status-phase">{texts.TITLE}</h3>
          {message}
        </Panel.Body>
      </Panel>
    </div>
  );
}
