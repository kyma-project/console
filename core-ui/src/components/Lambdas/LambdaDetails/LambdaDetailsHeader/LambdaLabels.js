import React from 'react';

import { Labels } from 'react-shared';

import EditLambdaLabelsModal from './EditLambdaLabelsModal';

import { LAMBDA_DETAILS } from 'components/Lambdas/constants';

export default function LambdaLabels({ lambda }) {
  return (
    <div className="lambda-details-header__labels">
      <h3 className="lambda-details-header__labels-title fd-has-color-text-4">
        <span>{LAMBDA_DETAILS.LABELS.TITLE}</span>
        <EditLambdaLabelsModal lambda={lambda} />
      </h3>
      <div className="lambda-details-header__labels-content fd-has-color-text-1">
        <Labels labels={lambda.labels} />
      </div>
    </div>
  );
}
