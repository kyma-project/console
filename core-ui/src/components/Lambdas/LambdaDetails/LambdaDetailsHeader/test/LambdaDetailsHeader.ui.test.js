import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import LambdaDetailsHeader from '../LambdaDetailsHeader';

describe('LambdaDetailsHeader UI', () => {
  it('Renders with minimal props', () => {
    const component = renderer.create(
      <LambdaDetailsHeader
        lambda={{ name: 'testlambda', namepace: 'testnamespace' }}
        handleUpdate={() => {}}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Handles save button click', () => {
    const handleUpdate = jest.fn();

    const component = mount(
      <LambdaDetailsHeader
        lambda={{ name: 'testlambda', namepace: 'testnamespace' }}
        handleUpdate={handleUpdate}
      />,
    );
    const deleteButton = component.find(
      '[data-test-id="lambda-save-button"] button',
    );
    deleteButton.simulate('click');

    expect(handleUpdate).toHaveBeenCalledTimes(1);
  });
});
