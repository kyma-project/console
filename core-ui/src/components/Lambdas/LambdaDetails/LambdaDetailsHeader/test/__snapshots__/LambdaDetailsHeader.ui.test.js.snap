// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaDetailsHeader UI Renders with minimal props 1`] = `
<header
  className="fd-has-background-color-background-2"
>
  <section
    className="fd-has-padding-regular fd-has-padding-bottom-none action-bar-wrapper"
  >
    <section>
      <ul
        className="fd-breadcrumb"
      >
        <li
          className="fd-breadcrumb__item"
          onClick={[Function]}
        >
          <a
            className="fd-breadcrumb__link"
            href="#"
          >
            Lambdas
          </a>
        </li>
        <li
          className="fd-breadcrumb__item"
        />
      </ul>
      <div
        className="fd-action-bar__header"
      >
        <h3
          className="fd-action-bar__title"
        >
          testlambda
        </h3>
      </div>
    </section>
    <div
      className="fd-action-bar__actions"
    >
      <button
        className="fd-button"
        data-test-id="lambda-save-button"
        onClick={[Function]}
      >
        Save
      </button>
      <button
        className="fd-button--light fd-button--negative"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </section>
</header>
`;
